client/code_generator/generator.go

```diff
@@ -1,10 +1,295 @@
+	hasEncounteredNonBarline := false
+	encounterBarline := func() {
+		if hasEncounteredNonBarline {
+			g.addToken(parser.Tie, "~")
+		}
+		hasEncounteredNonBarline = true
+	}
+
+	for _, component := range duration.Components {
+		switch value := component.(type) {
+		case model.Barline:
+			g.generateBarline(value)
+		case model.NoteLength:
+			encounterBarline()
+
+			noteLength := formatFloat(value.Denominator) +
+				strings.Repeat(".", int(value.Dots))
+			g.addToken(parser.NoteLength, noteLength)
+		case model.NoteLengthMs:
+			encounterBarline()
+
+			noteLengthMs := formatFloat(value.Quantity) + "ms"
+			g.addToken(parser.NoteLengthMs, noteLengthMs)
+		default:
+			// NoteLengthBeats is only generated from Lisp
+			// Duration cannot be nested in Alda
+			g.logError(component)
+		}
+	}
+}
+
+func (g *generator) generateNote(note model.Note) {
+	g.generatePitch(note.Pitch)
+	g.generateDuration(note.Duration)
+	if note.Slurred {
+		g.addToken(parser.Tie, "~")
+	}
+}
+
+func (g *generator) generateRest(rest model.Rest) {
+	g.addToken(parser.RestLetter, "r")
+	g.generateDuration(rest.Duration)
+}
+
+func (g *generator) generatePartDeclaration(decl model.PartDeclaration) {
+	for i, name := range decl.Names {
+		if i > 0 {
+			g.addToken(parser.Separator, "/")
+		}
+		g.addToken(parser.Name, name)
+	}
+
+	if decl.Alias != "" {
+		g.addToken(parser.Alias, "\"" + decl.Alias + "\"")
+	}
+
+	g.addToken(parser.Colon, ":")
+}
+
+func (g *generator) generateRepeat(repeat model.Repeat) {
+	g.generateTokens([]model.ScoreUpdate{repeat.Event})
+	g.addToken(parser.Repeat, "*" + strconv.Itoa(int(repeat.Times)))
+}
+
+func (g *generator) generateOnRepetitions(or model.OnRepetitions) {
+	g.generateTokens([]model.ScoreUpdate{or.Event})
+
+	repetitionsText := "'"
+	for i, repetition := range or.Repetitions {
+		if i > 0 {
+			repetitionsText += ","
+		}
+		if repetition.First == repetition.Last {
+			repetitionsText += strconv.Itoa(int(repetition.First))
+		} else {
+			repetitionsText += strconv.Itoa(int(repetition.First))
+			repetitionsText += "-"
+			repetitionsText += strconv.Itoa(int(repetition.Last))
+		}
+	}
+
+	g.addToken(parser.Repetitions, repetitionsText)
+}
+
+func (g *generator) generateVariableDeclaration(vd model.VariableDefinition) {
+	g.addToken(parser.Name, vd.VariableName)
+	g.addToken(parser.Equals, "=")
+	g.generateTokens(vd.Events)
+}
+
+func (g *generator) generateVariableReference(vr model.VariableReference) {
+	g.addToken(parser.Name, vr.VariableName)
+}
+
+func (g *generator) generateVoiceMarker(vm model.VoiceMarker) {
+	g.addToken(
+		parser.VoiceMarker,
+		"V" + strconv.Itoa(int(vm.VoiceNumber)) + ":",
+	)
+}
+
+func (g *generator) generateVoiceGroupEndMarker(_ model.VoiceGroupEndMarker) {
+	g.addToken(parser.VoiceMarker, "V0:")
+}
+
+func (g *generator) generateTokens(scoreUpdates []model.ScoreUpdate) {
+	for _, update := range scoreUpdates {
+		switch value := update.(type) {
```

Comment:
I was thinking that this could be a part of the ScoreUpdate interface, something like this:

```go
type ScoreUpdate interface {
	...

	// ASTNode returns an AST node representation of the ScoreUpdate.
	ASTNode() ast.Node
}

func (vm model.VoiceMarker) ASTNode() ast.Node {
	return ast.Node{...}
}

func (g *generator) generateAST(scoreUpdates []model.ScoreUpdate) {
	for _, update := range scoreUpdates {
		g.addASTNode(update.ASTNode())
	}
}
```

I think this would be nice because the compiler would force us to implement the ASTNode method for every new kind of ScoreUpdate that we introduce in the future.

(This is all a bit speculative, of course, because we haven't done the refactor to have the parser produce an AST yet.)

Reply:
Right so in my initial research, this wasn't a possibility. This is because Golang prohibits cyclic module dependencies. So `parser` already needs `model` obviously, and this would require `model` need `parser`. 

If we can find some way to circumvent this, it'll allow us to add the generation directly to the `ScoreUpdate` interface. If we directly use Golang's `ast.Node` that would work. If we want to define our own AST node, we would have to think of some code organization that let's us introduce a dependency within `model` that doesn't violate Go's rules. 

Either way I mostly opened these PRs as an initial guess of my ideas. It is very easy to refactor the logic here into interface implementations within `model`.

Reply:
I had that same thought. I was thinking that we could put our AST related code in a separate module from parser and model. I wasn't aware that there even was a Go AST package! :joy:  It looks like [ast.Node](https://pkg.go.dev/go/ast#Node) is not necessarily what we want, it's just an interface with a couple of functions related to the node's position in the text.

I'm still up for taking a crack at that refactor myself, BTW. Just a matter of finding some time for it!

Reply:
Haha ok. Sounds about right. I thought Go's build-in AST would not be what we wanted. A new package seems correct.